aya/src/maps/ringbuf.rs

```diff
@@ -0,0 +1,355 @@
+        if map_type != BPF_MAP_TYPE_RINGBUF as u32 {
+            return Err(MapError::InvalidMapType { map_type }.into());
+        }
+
+        // Determine page_size, map_fd, and set mask to map size - 1
+        let page_size = unsafe { sysconf(_SC_PAGESIZE) } as usize;
+        let map_fd = map.fd_or_err().map_err(RingBufferError::from)?;
+        let mask = (map.obj.def.max_entries - 1) as usize;
+
+        // Map writable consumer page
+        let consumer_page = unsafe {
+            mmap(
+                ptr::null_mut(),
+                page_size,
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED,
+                map_fd,
+                0,
+            )
+        };
+        if consumer_page == MAP_FAILED {
+            return Err(RingBufferError::ConsumerMMapError {
+                io_error: io::Error::last_os_error(),
+            });
+        }
+
+        // From kernel/bpf/ringbuf.c:
+        // Each data page is mapped twice to allow "virtual"
+        // continuous read of samples wrapping around the end of ring
+        // buffer area:
+        // ------------------------------------------------------
+        // | meta pages |  real data pages  |  same data pages  |
+        // ------------------------------------------------------
+        // |            | 1 2 3 4 5 6 7 8 9 | 1 2 3 4 5 6 7 8 9 |
+        // ------------------------------------------------------
+        // |            | TA             DA | TA             DA |
+        // ------------------------------------------------------
+        //                               ^^^^^^^
+        //                                  |
+        // Here, no need to worry about special handling of wrapped-around
+        // data due to double-mapped data pages. This works both in kernel and
+        // when mmap()'ed in user-space, simplifying both kernel and
+        // user-space implementations significantly.
+        let producer_pages = unsafe {
+            mmap(
+                ptr::null_mut(),
+                page_size + 2 * (mask + 1),
+                PROT_READ,
+                MAP_SHARED,
+                map_fd,
+                page_size as i64,
+            )
+        };
+        if producer_pages == MAP_FAILED {
+            return Err(RingBufferError::ProducerMMapError {
+                io_error: io::Error::last_os_error(),
+            });
+        }
+
+        Ok(RingBuf {
+            _map: Arc::new(map),
+            map_fd,
+            data_ptr: unsafe { (producer_pages as *mut u8).add(page_size) },
+            consumer_pos_ptr: consumer_page as *mut _,
+            producer_pos_ptr: producer_pages as *mut _,
+            page_size,
+            mask,
+        })
+    }
+
+    /// Retrieve an event from the ring, pass it to the callback, mark it as consumed, then repeat.
+    /// Returns when there's no more events.
+    pub fn process_ring(&mut self, callback: &mut impl FnMut(&[u8])) {
```

Comment:
Have we considered returning an iterator that consumes events instead of having a callback?

Reply:
We might be able to do that, but not with an `Iterator`, because then we get into the streaming iterator and self reference hell.

Return a non-Iterator handle struct that mutably borrows the RingBuf will probably work, and that can be used easily with a while loop like `while let Some(entry) = ringbuf.next()`. The reader pointer is incremented on `Drop`. What do you think?

Reply:
It's not immediately clear to me why it would be self referential?

Reply:
Oops, it's not a self reference, but since we can't use lifetime without GATs, the entry returned from the iterator will need something like `Rc` that references the main `RingBuf` to properly work.